src/main/java/duke/Duke.java

```diff
@@ -0,0 +1,267 @@
+                addTask(deadline, false);
+                break;
+            case "E":
+                Event event = new Event(words[2], words[1].equals("1"), words[3]);
+                addTask(event, false);
+                break;
+            default:
+                System.out.println("Error reading data from file: Invalid format");
+                break;
+            }
+        }
+    }
+
+    public  static  void todo(String line, String[] words) {
+        if (words.length >= 2) {
+            String task = line.substring(line.indexOf(' ') + 1);
+            Todo todo = new Todo(task, false);
+            addTask(todo, true);
+        } else {
+            System.out.println("Usage: todo <task>");
+        }
+    }
+
+    public static void deadline(String line) {
+        try {
+            String task = line.substring(line.indexOf(' ')+1, line.indexOf("/by")-1);
+            String date = line.substring(line.indexOf("/by")+4);
+            Deadline deadline = new Deadline(task, false, date);
+            addTask(deadline, true);
+        } catch (StringIndexOutOfBoundsException stringIndexOutOfBoundsException) {
+            System.out.println("Usage: deadline <task> /by <date>");
+        }
+    }
+
+    public static void event(String line) {
+        try {
+            String task = line.substring(line.indexOf(' ')+1, line.indexOf("/at")-1);
+            String date = line.substring(line.indexOf("/at")+4);
+            Event event = new Event(task, false, date);
+            addTask(event, true);
+        } catch (StringIndexOutOfBoundsException stringIndexOutOfBoundsException) {
+            System.out.println("Usage: event <task> /at <date>");
+        }
+    }
+
+    public static void list() throws ArrayEmptyException {
+        if (taskList.size() > 0) {
+            System.out.println("Here are the tasks in your list:");
+            for (int i = 0; i < taskList.size(); i++) {
+                System.out.printf("%d. %s\n", i + 1, taskList.get(i).print());
+            }
+        } else {
+            throw new ArrayEmptyException();
+        }
+    }
+
+    public static void mark(boolean isMark, String[] words) {
+        if (words.length == 2) {
+            try {
+                int d = Integer.parseInt(words[1]);
+                if (d > 0 && d <= taskList.size()) { // Valid
+                    if (taskList.get(d-1).getStatus()) { // Task is done
+                        if (isMark) {
+                            System.out.printf("This task is already done:\n" + "%s\n",
+                                    taskList.get(d - 1).print());
+                        } else {
+                            taskList.get(d - 1).setDone(false);
+                            System.out.printf("OK, I've marked this task as not done yet:\n" + "%s\n",
+                                    taskList.get(d - 1).print());
+                        }
```

Comment:
For this section of your code, consider abstracting out some sections of your code to maintain a high level of code readibility through reducing deeply nested code.